<!doctype html>
<html>
<head>
    <meta charset="utf-8"></meta>
    <title>BioThings Explorer Tutorial</title>
    <style type="text/css">
    	h2, h3 {
    		margin-left: 10%
    	}
    	img{
    		display: inline-block;
    		vertical-align: top;
    		width: 50%;
    		margin-left: 10%;
    		border-width: 5px;
    		border-color: blue;
    	}
    	p {
    		display: inline-block;
    		vertical-align: top;
    		width: 25%;
    		margin-right: 5%;
    		margin-left: 5%;
    	}
    	table, th, td {
    		border: 1px solid black;
		}
		table{
			margin-left: 10%;
			margin-right: 10%;
		}

    </style>
</head>
<body>
  <div class="back">
  	<form action="index.html">
  	  <input type="submit" value="Go to index" />
  	</form>
  </div>
  <div class="title">
    <h1 align="center"> Tutorial For BioThings Explorer </h3>
  </div>
  <div class="intro">
  	<h2> Introduction </h4>
  	<p>BioThings Explorer is an intuitive web-based graphical user interface for <a href="https://pypi.python.org/pypi/biothings-explorer/0.1.0">biothings-explorer Python Client</a> implemented using Cytoscape.js library, empowering users to explore the relationship between different biological entities through the vast amount of biological data provided by various API resources in a visually organized manner.</p>
  	<p>In this tutorial, we will walk you through a workflow, from gene -> variant -> drug, covering mulitple features in BioThings Explorer, including explore relationship, filtering results, changing layout, and action logs.
  </div>
  <hr>
  <br>
  <div class="main">
  	<h2> Tour </h2>
  	<div class="step1">
	  <h3> Step 1: Specify Input </h4>
	  <img src="../static/img/tutorial_input.jpg" alt="Input" style="width:500px;height:350px;">
	  <p> First, start by specifying the <b>ID TYPE</b> and <b>ID</b>. <br><br>Currently accepted <b>ID TYPEs</b> in BioThings Explorer includes entrez_gene_id, ensembl_gene_id, wikipathways_ids, etc. For a complete list of ID types accepted and examples of IDs, please refer to <a href="#bioentity">ID TYPE</a> section. <br><br>For demonstration, we will use 'entrez_gene_id' 1136 as a start point. <br><br>After filling 'ID TYPE' and 'ID' field, click <b>'Submit'</b>, and a node for the ID you specified will display on the 'Visualization' section.</p>
	</div>
	<br>
	<div class="aligncenter" style="height:0;border-top:2px dashed #FF00FF;font-size:0;">-</div>

	<br>
  	<div class="step2">
  	  <div>
		  <h3> Step 2: Click on Biological Entity node to find linked APIs </h4>
		  <img src="../static/img/tutorial_node.jpg" alt="Input" style="width:500px;height:350px;">
		  <p> Here, a red node is been displayed in the 'Visualization' box representing 'entrez_gene_id' 1136.</p>
	  </div>
	  <br>
	  <div class="aligncenter" style="height:0;border-top:2px dashed #FF00FF;font-size:0;">-</div>
	  <br>
	  <div>
		  <img src="../static/img/tutorial_click.jpg" alt="bioentity" style="width:500px;height:350px;">
		  <p> Now, click on the red node, it will display all avialable API resources which accepts 'entrez_gene_id' as an input. Currently, we have 3 APIs in BioThings Explorers which could be linked from 'entrez_gene_id', e.g. MyGene.info, MyVariant.info and BioLink.
		  <br><br> For a complete list of API resources included in BioThings Explorer, please check the <a href="#api">API Resource</a> section.</p>
	  </div>
	</div>
	<br>
	<div class="aligncenter" style="height:0;border-top:2px dashed #FF00FF;font-size:0;">-</div>
	<br>
	<div class="step3">
	  <div>
	      <h3> Step 3: Click on API node to find linked Biological Entities </h3>
	      <img src="../static/img/tutorial_click_variant.jpg" alt="variant" style="width:500px;height:350px;">
	      <p> Here, we click on the node for <a href="https://myvariant.info">myvariant.info</a>, which is an API for variant annotation information. <br><br>The system automatically make API calls to MyVariant.info and returns all hgvs_ids in MyVariant.info which are related to 'entrez_gene_id' 1136 , which is the initial input.<br><br>The results are listed in the left 'Results' box. You can browse all results using the pagination bar below.
	      </p>
	  </div>
	</div>
	<br>
	<div class="aligncenter" style="height:0;border-top:2px dashed #FF00FF;font-size:0;">-</div>
	<br>
	<div class="step4">
	  <div>
	      <h3> Step 3: Click interested result to add to graph </h3>
	      <img src="../static/img/tutorial_click_result.jpg" alt="variant" style="width:500px;height:350px;">
	      <p> Here, we click on the one of the result returned from MyVariant.info, which is 'chr15:g.78888992G>A'. It is then added to the 'Visualization' box. And you can click on that 'chr15:g.78888992G>A' node to explore further about all APIs and biogical entities which could be linked from this variant.
	      </p>
	  </div>
	</div>
	<br>
	<div class="aligncenter" style="height:0;border-top:2px dashed #FF00FF;font-size:0;">-</div>
	<br>
	<div class="step4">
	  <div>
	      <h3> Step 4: Filter </h3>
	      <img src="../static/img/tutorial_filter.jpg" alt="variant" style="width:500px;height:350px;">
	      <p> In case you want to filter the results returned, you could use the 'Filters' box on the left column. Here we filter for all variants which has allele frequency recorded lower than 0.15 in 1000 Genome Project. You can add multiple filters by clicking on the 'Add' button in the 'Filter' box. When finished, click 'Update' button in the 'Filter' box to update your search results. 
	      </p>
	  </div>
	</div>
	<br>
	<div class="aligncenter" style="height:0;border-top:2px dashed #FF00FF;font-size:0;">-</div>
	<br>
	  <div>
	      <h3> Step 5: Check log file </h3>
	      <img src="../static/img/tutorial_log.jpg" alt="variant" style="width:500px;height:350px;">
	      <p> There is box called 'Action Logs' on the right column, which records all actions which you have performed on BioThings Explorer. From there, you could get all API call information for further analysis.  
	      </p>
	  </div>
	</div>
	<br>
	<div class="aligncenter" style="height:0;border-top:2px dashed #FF00FF;font-size:0;">-</div>
	<br>
	</div>
	  <div>
	      <h3> Step 6: Change layout </h3>
	      <img src="../static/img/tutorial_layout.jpg" alt="variant" style="width:500px;height:350px;">
	      <p> There is box called 'Layout' on the left column, which you could use to change 'layout' of the graph displayed on the 'Visualization' box. Current options for layout includes: breadthfirst, grid, circle, concentric and cose.</p>  
	      </p>
	  </div>
	</div>
	<br>
	<hr>
	<br>
  </div>
  <div class="meta">
  	<div id="api">
  	  <h2> Available API Resources in BioThings Explorer</h2>
  	  <ul>
  	    <li><a href="https://myvariant.info">myvariant.info</a></li>
  	    <li><a href="https://mygene.info">mygene.info</a></li>
  	    <li><a href="https://mychem.info">mychem.info</a></li>
  	    <li><a href="http://dgidb.genome.wustl.edu/api">DGIdb</a></li>
  	    <li><a href="http://lynx.ci.uchicago.edu/">LYNX</a></li>
  	    <li><a href="https://api.monarchinitiative.org/api/">BioLink</a></li>
  	    <li><a href="https://reactome.org">Reactome</a></li>
  	    <li><a href="https://open.fda.gov/api/">openFDA</a></li>
  	    <li><a href="http://www.omicsdi.org/">OmicsDI</a></li>
  	    <li><a href="https://civic.genome.wustl.edu/">CIViC</a></li>
  	    <li><a href="https://clue.io/api">CLUE API</a></li>
  	  </ul>
  	</div>
 	<br>
	<div class="aligncenter" style="height:0;border-top:2px dashed #FF00FF;font-size:0;">-</div>
	<br>
  	<div id="bioentity">
  	  <h2> Available Biological Entities in BioThings Explorer</h2>
		<table style="width:70%">
		  <tr>
		    <th>ID</th>
		    <th>Example</th> 
		  </tr>
		  <tr>
		    <td>ensembl_gene_id</td>
		    <td>ENSG00000139618</td> 
		  </tr>
		  <tr>
		    <td>ensembl_protein_id</td>
		    <td>ENSP0000012345</td> 
		  </tr>
		  <tr>
		    <td>ensembl_transcript_id</td>
		    <td>ENST0000012345</td> 
		  </tr>
		  <tr>
		    <td>gene_ontology_id</td>
		    <td>GO:1234556</td> 
		  </tr>
		  <tr>
		    <td>interpro_id</td>
		    <td>IPR011009</td> 
		  </tr>
		  <tr>
		    <td>entrez_gene_id</td>
		    <td>1017</td> 
		  </tr>
		  <tr>
		    <td>hgnc_gene_symbol</td>
		    <td>CDK7</td> 
		  </tr>
		  <tr>
		    <td>biocarta_id</td>
		    <td>cellcyclepathway</td> 
		  </tr>
		  <tr>
		    <td>dbsnp_id</td>
		    <td>rs123456</td> 
		  </tr>
		  <tr>
		    <td>pubchem_id</td>
		    <td>100101</td> 
		  </tr>
		  <tr>
		    <td>pubmed_id</td>
		    <td>16333295</td> 
		  </tr>
		  <tr>
		    <td>uniprot_id</td>
		    <td>P62158</td> 
		  </tr>
		  <tr>
		    <td>wikipathway_id</td>
		    <td>WP100</td> 
		  </tr>
		  <tr>
		    <td>clinicaltrial_id</td>
		    <td>NCT01314001</td> 
		  </tr>
		  <tr>
		    <td>symptom</td>
		    <td>Microcephanly</td> 
		  </tr>
		  <tr>
		    <td>diseaseontology_id</td>
		    <td>DOID:1089</td> 
		  </tr>
		  <tr>
		    <td>disease_name</td>
		    <td>Intellectual disability</td> 
		  </tr>
		  <tr>
		    <td>reactome_id</td>
		    <td>R-HSA-199420</td> 
		  </tr>
		  <tr>
		    <td>pathway_name</td>
		    <td>Deubiquitination</td> 
		  </tr>
		  <tr>
		    <td>drug_effect</td>
		    <td>Renal impairment</td> 
		  </tr>
		  <tr>
		    <td>pharmacology_class</td>
		    <td>Nonsteroidal Anti-inflammatory Compounds [Chemical/Ingredient]</td> 
		  </tr>
		  <tr>
		    <td>omim_term</td>
		    <td>Myelokathexis</td> 
		  </tr>
		  <tr>
		    <td>chembl_id</td>
		    <td>CHEMBL1949675</td> 
		  </tr>
		  <tr>
		    <td>pdb_id</td>
		    <td>2K03</td> 
		  </tr>
		  <tr>
		    <td>mgi_term</td>
		    <td>endocrine/exocrine gland phenotype</td> 
		  </tr>
		  <tr>
		    <td>mouse_gene_id</td>
		    <td>1860276</td> 
		  </tr>
		  <tr>
		    <td>civic_gene_id</td>
		    <td>6</td> 
		  </tr>
		  <tr>
		    <td>civic_variant_id</td>
		    <td>6</td> 
		  </tr>
		  <tr>
		    <td>civic_evidence_id</td>
		    <td>6</td> 
		  </tr>
		  <tr>
		    <td>so_id</td>
		    <td>SO:0001121</td> 
		  </tr>
		  <tr>
		    <td>rxcui</td>
		    <td>603952</td> 
		  </tr>
		  <tr>
		    <td>hgvs_id</td>
		    <td>chr6:123456G>A</td> 
		  </tr>
		</table>
  	</div>

</body>

